File: 1984-maximum-distance-between-a-pair-of-values/maximum-distance-between-a-pair-of-values.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxDistance(vector<int>& nums1, vector<int>& nums2) {
        int i = 0, j = 0;
        int maxDist = 0;
        int n1 = nums1.size();
        int n2 = nums2.size();

        while (i < n1 && j < n2) {
            // Check if the current pair is valid
            if (nums1[i] <= nums2[j]) {
                maxDist = max(maxDist, j - i);
                // Try to increase distance by moving j
                j++;
            } else {
                // Current nums1[i] is too large, move to a smaller value
                i++;
            }
        }
        
        return maxDist;
    }
};

File: 1984-maximum-distance-between-a-pair-of-values/maximum-distance-between-a-pair-of-values.java
class Solution {
    public int maxDistance(int[] nums1, int[] nums2) {
        int i = 0;
        int j = 0;
        int maxDist = 0;
        int n1 = nums1.length;
        int n2 = nums2.length;

        while (i < n1 && j < n2) {
            if (nums1[i] <= nums2[j]) {
                // Update max distance if valid
                maxDist = Math.max(maxDist, j - i);
                // Look for a further j
                j++;
            } else {
                // Move i forward to decrease nums1[i]
                i++;
            }
        }

        return maxDist;
    }
}